lib/rules/media-feature-name-unit-allowed-list/index.js

```diff
@@ -0,0 +1,94 @@
+'use strict';
+
+const report = require('../../utils/report');
+const ruleMessages = require('../../utils/ruleMessages');
+const validateOptions = require('../../utils/validateOptions');
+const { TokenType } = require('@csstools/css-tokenizer');
+const { isTokenNode } = require('@csstools/css-parser-algorithms');
+const {
+	isMediaFeaturePlain,
+	isMediaFeatureRange,
+	parse,
+} = require('@csstools/media-query-list-parser');
+const validateObjectWithArrayProps = require('../../utils/validateObjectWithArrayProps');
+const { isString } = require('../../utils/validateTypes');
+const matchesStringOrRegExp = require('../../utils/matchesStringOrRegExp');
+const flattenArray = require('../../utils/flattenArray');
+
+const ruleName = 'media-feature-name-unit-allowed-list';
+
+const messages = ruleMessages(ruleName, {
+	rejected: (unit, name) => `Unexpected unit "${unit}" for name "${name}"`,
+});
+
+const meta = {
+	url: 'https://stylelint.io/user-guide/rules/media-feature-name-unit-allowed-list',
+};
+
+/** @type {import('stylelint').Rule} */
+const rule = (primary) => {
+	return (root, result) => {
+		const validOptions = validateOptions(result, ruleName, {
+			actual: primary,
+			possible: [validateObjectWithArrayProps(isString)],
+		});
+
+		if (!validOptions) {
+			return;
+		}
+
+		root.walkAtRules(/^media$/i, (atRule) => {
+			const mediaQueryList = parse(atRule.params);
+
+			mediaQueryList.forEach((mediaQuery) => {
+				mediaQuery.walk((entry) => {
+					if (!isMediaFeaturePlain(entry.node) && !isMediaFeatureRange(entry.node)) {
+						return;
+					}
+
+					const name = entry.node.name.toString();
+
+					const nameKey = Object.keys(primary).find((nameIdentifier) =>
+						matchesStringOrRegExp(name, nameIdentifier),
+					);
+
+					if (!nameKey) {
+						return;
+					}
+
+					const nameList = flattenArray(primary[nameKey]);
+
+					if (!nameList) {
+						return;
+					}
+
+					entry.node.walk((childEntry) => {
+						if (!isTokenNode(childEntry.node) || childEntry.node.value[0] !== TokenType.Dimension) {
+							return;
+						}
+
+						const unit = childEntry.node.value[4].unit;
+
+						if (nameList.includes(unit.toLowerCase())) {
+							return;
+						}
+
+						report({
+							message: messages.rejected(unit, name),
+							node: atRule,
+							index: childEntry.node.value[2],
+							endIndex: childEntry.node.value[3],
```

Comment:
When I wrote this bit I didn't know what these values should be :)

After converting an existing rule I found that this works :

```
const index = atRuleParamIndex(atRule) + nameToken[2];
const endIndex = atRuleParamIndex(atRule) + nameToken[3] + 1;
```